lib/stellar/sep10.rb

```diff
@@ -0,0 +1,353 @@
+      server: Stellar::KeyPair
+    ] => nil)
+    # Verifies if a transaction is a valid per SEP-10 challenge transaction, if the validation 
+    # fails, an exception will be thrown.
+    #
+    # This function performs the following checks:
+    #   1. verify that transaction sequenceNumber is equal to zero;
+    #   2. verify that transaction source account is equal to the server's signing key;
+    #   3. verify that transaction has time bounds set, and that current time is between the minimum and maximum bounds;
+    #   4. verify that transaction contains a single Manage Data operation and it's source account is not null;
+    #   5. verify that transaction envelope has a correct signature by server's signing key;
+    #   6. verify that transaction envelope has a correct signature by the operation's source account;
+    #
+    # @param challenge_transaction [String] SEP0010 transaction challenge transaction in base64.
+    # @param server [Stellar::KeyPair] keypair for server's account.
+    #
+    # Raises a InvalidSep10ChallengeError if the validation fails
+    def self.verify_challenge_transaction(
+      challenge_transaction: String, server: Stellar::KeyPair
+    )
+      transaction_envelope, client_address = read_challenge_tx(
+          challenge: challenge_transaction, server: server
+      )
+      client_keypair = Stellar::KeyPair.from_address(client_address)
+      if !verify_tx_signed_by(transaction_envelope: transaction_envelope, keypair: client_keypair)
+        raise InvalidSep10ChallengeError.new(
+            "Transaction not signed by client: %s" % [client_keypair.address]
+        )
+      end
+    end
+
+    Contract(C::KeywordArgs[
+      transaction_envelope: Stellar::TransactionEnvelope, 
+      signers: ArrayOf[Stellar::AccountSigner]
+    ] => ArrayOf[Stellar::AccountSigner])
+    # Checks if a transaction has been signed by one or more of the signers, 
+    # returning a list of unique signers that were found to have signed the transaction.
+    #
+    # @param transaction_envelope [Stellar::TransactionEnvelope] SEP0010 transaction challenge transaction envelope.
+    # @param signers [ArrayOf[Stellar::AccountSigner]] The signers of client account.
+    #
+    # @return [ArrayOf[Stellar::AccountSigner]]
+    def self.verify_transaction_signatures(
+      transaction_envelope:,
+      signers:
+    )
+      signatures = transaction_envelope.signatures
+      if signatures.empty?
+        raise InvalidSep10ChallengeError.new("Transaction has no signatures.")
+      end
+
+      signers_seen = Set.new
+      signers_found = Array.new
+      signers.each do |signer|
+        if signers_seen.include?(signer.address)
+          next
+        end
+        signers_seen.add(signer.address)
+        kp = Stellar::KeyPair.from_address(signer.address)
+        if verify_tx_signed_by(transaction_envelope: transaction_envelope, keypair: kp)
+          signers_found.push(signer)
+        end
+      end
```

Comment:
This code matches signers to signatures one-by-one. It ensures that a signer isn't consumed more than once, but it doesn't ensure a signature is consumed more than once. We should match all signatures only once. This codes makes the assumption that the algorithm used for signing will never produce the same signature for two signers. If that ever happened you'd count a single signature twice. I don't know of a case where this would happen with ed25519, but we shouldn't assume. The Go SDK and stellar-core use similar approaches to ensuring this doesn't happen.

See the [Go SDK](https://github.com/stellar/go/blob/c41b3e4309822abc5610230266fc51134ee853fb/txnbuild/transaction.go#L679-L702).